test/EVMHost.cpp

```diff
@@ -343,6 +354,8 @@ evmc::Result EVMHost::call(evmc_message const& _message) noexcept
 		code = accounts[message.code_address].code;
 
 	auto& destination = accounts[message.recipient];
+	// Mark account as created if it is a CREATE or CREATE2 tx
+	destination.isCreated = (message.kind == EVMC_CREATE || message.kind == EVMC_CREATE2);
```

Comment:
You probably mean call, not transaction here, right? A create transation would be one where the contract is passed in as the data.
```suggestion
	// Mark account as created if it is a CREATE or CREATE2 call
	destination.isCreated = (message.kind == EVMC_CREATE || message.kind == EVMC_CREATE2);
```

Reply:
Also, won't this reset the flag if after creation someone performs a different kind of call to this account?

Reply:
> You probably mean call, not transaction here, right? A create transation would be one where the contract is passed in as the data.

Yes!

> Also, won't this reset the flag if after creation someone performs a different kind of call to this account?

Yeah, but this is the idea, the flag should only be `true` on contract creation, because this is the only scenario that `selfdestruct` keep the old behavior, i.e. the creation and `selfdestruct` happens in the same tx.

Reply:
But what happens if I, say, create a contract, destroy it immediately and then low-level call the address. The call will not succeed but won't revert (I can just ignore the result), but it will still reset the flag.

Or another situation: create, destroy, create again at the same address, then call.

Reply:
> But what happens if I, say, create a contract, destroy it immediately and then low-level call the address. The call will not succeed but won't revert (I can just ignore the result), but it will still reset the flag.

hum...but if you do a low-level call it would still need to be of the kind create or create2 no? I don't know how the test suite handle this case. But I will add more tests here cover this scenario.

> Or another situation: create, destroy, create again at the same address, then call.

This should not be possible anymore according with EIP-6780. I.e. use `create2` to reploy a contract at the same address should revert for `EVM >= Cancun`. See [Security Considerations section](https://eips.ethereum.org/EIPS/eip-6780).

I have a [test](https://github.com/ethereum/solidity/pull/14785/files#diff-9c8c5fcfd95797c71afac511de2438d74fd58a32518409eda80818b462638a85R83) for it, but I didn't use a low-level call. So yeah, we may need more tests.

Reply:
> but if you do a low-level call it would still need to be of the kind create or create2 no?

No, why? You do create first, but then the control gets back to you. Then you make the contract self-destruct. And then you just use `<address>.call()` that has nothing to do with create.

> This should not be possible anymore according with EIP-6780.

Are you sure? It does not say it explicitly, but I think it refers to upgrading the contract in a separate transaction. I think you should be able to upgrade it while you're still inside the original transaction that created it.

Reply:
> > but if you do a low-level call it would still need to be of the kind create or create2 no?

> No, why? You do create first, but then the control gets back to you. Then you make the contract self-destruct. And then you just use `<address>.call()` that has nothing to do with create.

Oh, right. I guess I misunderstood your comment then. In this case, it would simply fail, as you said, but the flag would not be set because the last call is not a `create` or `create2`. See: https://github.com/ethereum/solidity/pull/14785/files#diff-27f9bec6af0022980abf546f2598f32da0e70a4085de33b3a86275c3516abf29 it is not exactly the scenario you describe, but it does create a contract using a low-level call to `create2`, then destroy it and try to recreate it. It works on EVM `<=shanghai` but it shouldn't work for EVM `>=cancun`.
 
> Are you sure? It does not say it explicitly, but I think it refers to upgrading the contract in a separate transaction. I think you should be able to upgrade it while you're still inside the original transaction that created it.

Yes, this is what I mean. If it is in the same transaction that creates the contract, then yes, it should preserve the old behavior. In the `Security Considerations section` of the EIP they have the following:

> The following applications of SELFDESTRUCT will be broken and applications that use it in this way are not safe anymore:
> 
> - Where CREATE2 is used to redeploy a contract in the same place in order to make a contract upgradable. This is not supported anymore and [ERC-2535](https://eips.ethereum.org/EIPS/eip-2535) or other types of proxy contracts should be used instead.
> 
> - Where a contract depended on burning Ether via a SELFDESTRUCT with the contract as beneficiary, in a contract not created within the same transaction.
>

Reply:
> But the flag would be `false` regardless, because the last call is not a `create` or `create2`.

So isn't this a problem? This flag is used to decode whether to erase the account or not and if we do it wrong, something should break.

But actually, I'm not sure I fully understand how you intend this to work. Initially I assumed that we must be processing this flag for every *call* performed during execution, because otherwise it would not make sense. E.g each time a contract calls an external function and also when that external function calls another external function, and so on. But looking at the sources, it's only processed each *message* (transaction). Which means basically only for the top-level call. If you have a transaction where multiple contracts are created  deep in the call chain, does this even properly account for that? You can't tell that based on whether the whole transaction is a contract creation or not.

Reply:
Well, now I'm not sure if it is a problem haha. My understanding of the EIP was that it should be in the context of the transaction, not every call.

See the text below extracted from https://eips.ethereum.org/EIPS/eip-6780:
> A contract is considered created at the beginning of a create transaction or when a CREATE series operation begins execution (CREATE, CREATE2, and other operations that deploy contracts in the future). If a balance exists at the contract’s new address it is still considered to be a contract creation


> If you have a transaction where multiple contracts are created deep in the call chain, does this even properly account for that?

Probably not in the way it's currently done, but we should test that. Aren't each call handled by `EVMHost::call` in the end?

Reply:
hum...but I guess this part: `OR when a CREATE series operation begins execution (CREATE, CREATE2, and other operations that deploy contracts in the future)` could mean exactly what you said, about a tx that could contain a bunch of contracts created deep in the call chain.

Reply:
Note that it says:

> at the beginning of a create transaction **or** when a CREATE series operation begins execution

i.e. it's both. A contract creation transaction does this but also a normal transaction that just happens to execute `CREATE` or `CREATE2` at some point.

> Probably not in the way it's currently done, but we should test that.

Yeah. And you can cover both in the same tests. Just run `deploy_and_terminate()` both in the constructor of the top-level contract and as a part of expectations. Could also be a separate test, but these tests are long, at least as they are now, so I'd try to reduce the boilerplate we need.

> Aren't each call handled by `EVMHost::call` in the end?

I don't think so. Note that its argument is a message. Also, it's our function, and it's `ExecutionFramework` that calls it. For it to run for every call we'd have to somehow pass it to evmone as a callback and I don't think we're doing that.

Reply:
> i.e. it's both. A contract creation transaction does this but also a normal transaction that just happens to execute `CREATE` or `CREATE2` at some point.

Yeah, after carefully rereading it, I also noticed that. Thanks for pointing it out. I've fixed the code and am adjusting the tests accordingly.

> Yeah. And you can cover both in the same tests. Just run `deploy_and_terminate()` both in the constructor of the top-level contract and as a part of expectations. Could also be a separate test, but these tests are long, at least as they are now, so I'd try to reduce the boilerplate we need.

Yes, I'll do that.

> > Aren't each call handled by `EVMHost::call` in the end?
> 
> I don't think so. Note that its argument is a message. Also, it's our function, and it's `ExecutionFramework` that calls it. For it to run for every call we'd have to somehow pass it to evmone as a callback and I don't think we're doing that.

It appears that it is invoked for every call, at least in my tests here. However, the flag was indeed reset in the wrong place. I'm now resetting it only when a new transaction frame is created.

Reply:
> It appears that it is invoked for every call

Hmm... ok, the comment over `EVMHost::call()` does say it gets triggered for internal calls. I guess it means both of them are treated uniformly. Good.